--- mdadm-3.2.2/mdadm.h.orig	2011-07-24 04:55:04.976315824 +0200
+++ mdadm-3.2.2/mdadm.h	2011-07-24 04:56:06.956183559 +0200
@@ -113,6 +113,10 @@
  * and there is no standard conversion function so... */
 /* And dietlibc doesn't think byteswap is ok, so.. */
 /*  #include <byteswap.h> */
+#undef bswap_16
+#undef bswap_32
+#undef bswap_64
+
 #define bswap_16(x) (((x) & 0x00ffU) << 8 | \
 		     ((x) & 0xff00U) >> 8)
 #define bswap_32(x) (((x) & 0x000000ffU) << 24 | \
