--- openldap-2.4.23/configure.in.bak	2012-06-26 15:38:17.467586187 +0200
+++ openldap-2.4.23/configure.in	2012-06-26 15:48:45.752811742 +0200
@@ -1887,7 +1887,7 @@
 	dnl $ol_cv_lib_db should be yes or -ldb
 	dnl (it could be no, but that would be an error
 	if test $ol_cv_lib_db != yes ; then
-		BDB_LIBS="$BDB_LIBS $ol_cv_lib_db"
+		BDB_LIBS="$BDB_LIBS $BDB_LIB_STATIC"
 	fi
 
 	SLAPD_LIBS="$SLAPD_LIBS \$(BDB_LIBS)"
