diff -Naur mdadm-3.2.2/.gitignore mdadm-6218489/.gitignore
--- mdadm-3.2.2/.gitignore	2011-04-24 14:23:10.000000000 +0200
+++ mdadm-6218489/.gitignore	2011-08-02 05:36:08.000000000 +0200
@@ -2,6 +2,7 @@
 /*.man
 /*-stamp
 /mdadm
+/mdadm.8
 /mdadm.udeb
 /mdmon
 /swap_super
diff -Naur mdadm-3.2.2/Create.c mdadm-6218489/Create.c
--- mdadm-3.2.2/Create.c	2011-06-16 08:53:22.000000000 +0200
+++ mdadm-6218489/Create.c	2011-08-02 05:36:08.000000000 +0200
@@ -856,15 +856,6 @@
 					/* getinfo_super might have lost these ... */
 					inf->disk.major = major(stb.st_rdev);
 					inf->disk.minor = minor(stb.st_rdev);
-					/* FIXME the following should not be needed
-					 * as getinfo_super is suppose to set
-					 * them.  However it doesn't for imsm,
-					 * so we have this hack for now
-					 */
-					if (st->ss == &super_imsm) {
-						inf->disk.number = dnum;
-						inf->disk.raid_disk = dnum;
-					}
 				}
 				break;
 			case 2:
diff -Naur mdadm-3.2.2/Detail.c mdadm-6218489/Detail.c
--- mdadm-3.2.2/Detail.c	2011-04-24 14:23:10.000000000 +0200
+++ mdadm-6218489/Detail.c	2011-08-02 05:36:08.000000000 +0200
@@ -372,11 +372,13 @@
 			else
 				st = ", degraded";
 
-			printf("          State : %s%s%s%s\n",
-			       (array.state&(1<<MD_SB_CLEAN))?"clean":"active",
-			       st,
-			       (!e || e->percent < 0) ? "" : sync_action[e->resync],
-			       larray_size ? "": ", Not Started");
+			printf("          State : %s%s%s%s%s%s \n",
+			       (array.state&(1<<MD_SB_CLEAN))?"clean":"active", st,
+			       (!e || (e->percent < 0 && e->percent != PROCESS_PENDING &&
+			       e->percent != PROCESS_DELAYED)) ? "" : sync_action[e->resync],
+			       larray_size ? "": ", Not Started",
+			       e->percent == PROCESS_DELAYED ? " (DELAYED)": "",
+			       e->percent == PROCESS_PENDING ? " (PENDING)": "");
 		}
 		if (array.raid_disks)
 			printf(" Active Devices : %d\n", array.active_disks);
@@ -416,10 +418,8 @@
 		}
 
 		if (e && e->percent >= 0) {
-			printf(" Re%s Status : %d%% complete\n",
-			       (st && st->sb && info->reshape_active)?
-			          "shape":"build",
-			       e->percent);
+			static char *sync_action[] = {"Rebuild", "Resync", "Reshape", "Check"};
+			printf(" %7s Status : %d%% complete\n", sync_action[e->resync], e->percent);
 			is_rebuilding = 1;
 		}
 		free_mdstat(ms);
@@ -430,12 +430,9 @@
 			printf("  Reshape pos'n : %llu%s\n", (unsigned long long) info->reshape_progress<<9,
 			       human_size((unsigned long long)info->reshape_progress<<9));
 #endif
-			if (info->delta_disks > 0)
+			if (info->delta_disks != 0)
 				printf("  Delta Devices : %d, (%d->%d)\n",
 				       info->delta_disks, array.raid_disks - info->delta_disks, array.raid_disks);
-			if (info->delta_disks < 0)
-				printf("  Delta Devices : %d, (%d->%d)\n",
-				       info->delta_disks, array.raid_disks, array.raid_disks + info->delta_disks);
 			if (info->new_level != array.level) {
 				char *c = map_num(pers, info->new_level);
 				printf("      New Level : %s\n", c?c:"-unknown-");
diff -Naur mdadm-3.2.2/Grow.c mdadm-6218489/Grow.c
--- mdadm-3.2.2/Grow.c	2011-06-17 07:15:50.000000000 +0200
+++ mdadm-6218489/Grow.c	2011-08-02 05:36:08.000000000 +0200
@@ -1268,7 +1268,7 @@
 
 	if (re->after.data_disks < re->before.data_disks &&
 	    get_linux_version() < 2006030)
-		return "reshape to fewer devices is not supported before 2.6.32 - sorry.";
+		return "reshape to fewer devices is not supported before 2.6.30 - sorry.";
 
 	re->backup_blocks = compute_backup_blocks(
 		info->new_chunk, info->array.chunk_size,
@@ -1494,7 +1494,7 @@
 			goto release;
 		}
 		if (assume_clean) {
-			/* This will fail on kernels newer than 2.6.40 unless
+			/* This will fail on kernels newer than 3.0 unless
 			 * a backport has been arranged.
 			 */
 			if (sra == NULL ||
@@ -2725,15 +2725,21 @@
 		int rv = -2;
 		tv.tv_sec = 10;
 		tv.tv_usec = 0;
-		while (fd >= 0 && rv < 0) {
+		while (fd >= 0 && rv < 0 && tv.tv_sec > 0) {
 			fd_set rfds;
 			FD_ZERO(&rfds);
 			FD_SET(fd, &rfds);
 			if (select(fd+1, NULL, NULL, &rfds, &tv) != 1)
 				break;
-			if (sysfs_fd_get_ll(fd, &completed) >= 0)
+			switch (sysfs_fd_get_ll(fd, &completed)) {
+			case 0:
 				/* all good again */
 				rv = 1;
+				break;
+			case -2: /* read error - abort */
+				tv.tv_sec = 0;
+				break;
+			}
 		}
 		if (fd >= 0)
 			close(fd);
diff -Naur mdadm-3.2.2/Kill.c mdadm-6218489/Kill.c
--- mdadm-3.2.2/Kill.c	2011-04-24 14:23:10.000000000 +0200
+++ mdadm-6218489/Kill.c	2011-08-02 05:36:08.000000000 +0200
@@ -59,6 +59,7 @@
 		close(fd);
 		return 2;
 	}
+	st->ignore_hw_compat = 1;
 	rv = st->ss->load_super(st, fd, dev);
 	if (force && rv >= 2)
 		rv = 0; /* ignore bad data in superblock */
diff -Naur mdadm-3.2.2/md.4 mdadm-6218489/md.4
--- mdadm-3.2.2/md.4	2011-04-24 14:23:10.000000000 +0200
+++ mdadm-6218489/md.4	2011-08-02 05:36:08.000000000 +0200
@@ -128,7 +128,7 @@
 From release 2.6.28, the
 .I md
 driver supports arrays with externally managed metadata.  That is,
-the metadata is not managed by the kernel by rather by a user-space
+the metadata is not managed by the kernel but rather by a user-space
 program which is external to the kernel.  This allows support for a
 variety of metadata formats without cluttering the kernel with lots of
 details.
@@ -468,7 +468,7 @@
 found that is a mismatch.
 .I md
 normally works in units much larger than a single sector and when it
-finds a mismatch, it does not determin exactly how many actual sectors were
+finds a mismatch, it does not determine exactly how many actual sectors were
 affected but simply adds the number of sectors in the IO unit that was
 used.  So a value of 128 could simply mean that a single 64KB check
 found an error (128 x 512bytes = 64KB).
@@ -661,7 +661,7 @@
 This is the partner of
 .B md/sync_speed_min
 and overrides
-.B /proc/sys/dev/raid/spool_limit_max
+.B /proc/sys/dev/raid/speed_limit_max
 described below.
 
 .TP
diff -Naur mdadm-3.2.2/mdadm.8.in mdadm-6218489/mdadm.8.in
--- mdadm-3.2.2/mdadm.8.in	2011-06-17 07:15:50.000000000 +0200
+++ mdadm-6218489/mdadm.8.in	2011-08-02 05:36:08.000000000 +0200
@@ -418,11 +418,22 @@
 A suffix of 'M' or 'G' can be given to indicate Megabytes or
 Gigabytes respectively.
 
+Sometimes a replacement drive can be a little smaller than the
+original drives though this should be minimised by IDEMA standards.
+Such a replacement drive will be rejected by
+.IR md .
+To guard against this it can be useful to set the initial size
+slightly smaller than the smaller device with the aim that it will
+still be larger than any replacement.
+
 This value can be set with
 .B \-\-grow
-for RAID level 1/4/5/6.  If the array was created with a size smaller
-than the currently active drives, the extra space can be accessed
-using
+for RAID level 1/4/5/6 though
+.B CONTAINER
+based arrays such as those with IMSM metadata may not be able to
+support this.
+If the array was created with a size smaller than the currently
+active drives, the extra space can be accessed using
 .BR \-\-grow .
 The size can be given as
 .B max
@@ -440,9 +451,10 @@
 .B "\-\-grow \-\-size="
 command.
 
-This value can not be used with
+This value cannot be used when creating a
 .B CONTAINER
-metadata such as DDF and IMSM.
+such as with DDF and IMSM metadata, though it perfectly valid when
+creating an array inside a container.
 
 .TP
 .BR \-Z ", " \-\-array\-size=
@@ -706,7 +718,7 @@
 When an array is resized to a larger size with
 .B "\-\-grow \-\-size="
 the new space is normally resynced in that same way that the whole
-array is resynced at creation.  From Linux version 2.6.40,
+array is resynced at creation.  From Linux version 3.0,
 .B \-\-assume\-clean
 can be used with that command to avoid the automatic resync.
 
@@ -811,6 +823,11 @@
 non-standard name.  Names that are not in 'standard' format are only
 allowed in "/dev/md/".
 
+This is meaningful with
+.B \-\-create
+or
+.BR \-\-build .
+
 .ig XX
 .\".TP
 .\".BR \-\-symlink = no
@@ -835,6 +852,28 @@
 .\"
 .XX
 
+.TP
+.BR \-a ", " "\-\-add"
+This option can be used in Grow mode in two cases.
+
+If the target array is a Linear array, then
+.B \-\-add
+can be used to add one or more devices to the array.  They
+are simply catenated on to the end of the array.  Once added, the
+devices cannot be removed.
+
+If the
+.B \-\-raid\-disks
+option is being used to increase the number of devices in an array,
+then
+.B \-\-add
+can be used to add some extra devices to be included in the array.
+In most cases this is not needed as the extra devices can be added as
+spares first, and then the number of raid-disks can be changed.
+However for RAID0, it is not possible to add spares.  So to increase
+the number of devices in a RAID0, it is necessary to set the new
+number of devices, and to add the new devices, in the same command.
+
 .SH For assemble:
 
 .TP
@@ -912,28 +951,6 @@
 See this option under Create and Build options.
 
 .TP
-.BR \-a ", " "\-\-add"
-This option can be used in Grow mode in two cases.
-
-If the target array is a Linear array, then
-.B \-\-add
-can be used to add one or more devices to the array.  They
-are simply catenated on to the end of the array.  Once added, the
-devices cannot be removed.
-
-If the
-.B \-\-raid\-disks
-option is being used to increase the number of devices in an array,
-then
-.B \-\-add
-can be used to add some extra devices to be included in the array.
-In most cases this is not needed as the extra devices can be added as
-spares first, and then the number of raid-disks can be changed.
-However for RAID0, it is not possible to add spares.  So to increase
-the number of devices in a RAID0, it is necessary to set the new
-number of devices, and to add the new devices, in the same command.
-
-.TP
 .BR \-b ", " \-\-bitmap=
 Specify the bitmap file that was given when the array was created.  If
 an array has an
@@ -1491,7 +1508,7 @@
 In the second usage example, all devices listed are treated as md
 devices and assembly is attempted.
 In the third (where no devices are listed) all md devices that are
-listed in the configuration file are assembled.  If not arrays are
+listed in the configuration file are assembled.  If no arrays are
 described by the configuration file, then any arrays that
 can be found on unused devices will be assembled.
 
@@ -1600,7 +1617,7 @@
 will first attempt to assemble all the arrays listed in the config
 file.
 
-In no array at listed in the config (other than those marked
+If no arrays are listed in the config (other than those marked
 .BR <ignore> )
 it will look through the available devices for possible arrays and
 will try to assemble anything that it finds.  Arrays which are tagged
@@ -2200,22 +2217,24 @@
 .IP \(bu 4
 increase or decrease the "raid\-devices" attribute of RAID0, RAID1, RAID4,
 RAID5, and RAID6.
-.IP \bu 4
+.IP \(bu 4
 change the chunk-size and layout of RAID0, RAID4, RAID5 and RAID6.
-.IP \bu 4
+.IP \(bu 4
 convert between RAID1 and RAID5, between RAID5 and RAID6, between
-RAID0, RAID5, and RAID5, and between RAID0 and RAID10 (in the near-2 mode).
+RAID0, RAID4, and RAID5, and between RAID0 and RAID10 (in the near-2 mode).
 .IP \(bu 4
 add a write-intent bitmap to any array which supports these bitmaps, or
 remove a write-intent bitmap from such an array.
 .PP
 
-Using GROW on containers is currently only support for Intel's IMSM
+Using GROW on containers is currently supported only for Intel's IMSM
 container format.  The number of devices in a container can be
 increased - which affects all arrays in the container - or an array
 in a container can be converted between levels where those levels are
 supported by the container, and the conversion is on of those listed
-above.
+above.  Resizing arrays in an IMSM container with
+.B "--grow --size"
+is not yet supported.
 
 Grow functionality (e.g. expand a number of raid devices) for Intel's
 IMSM container format has an experimental status. It is guarded by the
@@ -2250,7 +2269,7 @@
 are synchronised.
 
 Note that when an array changes size, any filesystem that may be
-stored in the array will not automatically grow for shrink to use or
+stored in the array will not automatically grow or shrink to use or
 vacate the space.  The
 filesystem will need to be explicitly told to use the extra space
 after growing, or to reduce its size
@@ -2259,7 +2278,7 @@
 
 Also the size of an array cannot be changed while it has an active
 bitmap.  If an array has a bitmap, it must be removed before the size
-can be changed. Once the change it complete a new bitmap can be created.
+can be changed. Once the change is complete a new bitmap can be created.
 
 .SS RAID\-DEVICES CHANGES
 
@@ -2435,8 +2454,8 @@
 line.
 
 .IP +
-Does the device have a valid md superblock.  If a specific metadata
-version is request with
+Does the device have a valid md superblock?  If a specific metadata
+version is requested with
 .B \-\-metadata
 or
 .B \-e
@@ -2467,6 +2486,7 @@
 current host, the device will be rejected.
 ..
 
+.PP
 .I mdadm
 keeps a list of arrays that it has partially assembled in
 .B /var/run/mdadm/map
@@ -2639,7 +2659,7 @@
 Any devices which are components of /dev/md4 will be marked as faulty
 and then remove from the array.
 
-.B "  mdadm --grow /dev/md4 --level=6 --backup-file=/root/backup-md4
+.B "  mdadm --grow /dev/md4 --level=6 --backup-file=/root/backup-md4"
 .br
 The array
 .B /dev/md4
@@ -2787,7 +2807,7 @@
 For further information on mdadm usage, MD and the various levels of
 RAID, see:
 .IP
-.B http://linux\-raid.osdl.org/
+.B http://raid.wiki.kernel.org/
 .PP
 (based upon Jakob \(/Ostergaard's Software\-RAID.HOWTO)
 .\".PP
diff -Naur mdadm-3.2.2/mdadm.h mdadm-6218489/mdadm.h
--- mdadm-3.2.2/mdadm.h	2011-06-14 04:50:01.000000000 +0200
+++ mdadm-6218489/mdadm.h	2011-08-02 05:36:08.000000000 +0200
@@ -1345,3 +1345,5 @@
 #define PATH_MAX	4096
 #endif
 
+#define PROCESS_DELAYED -2
+#define PROCESS_PENDING -3
diff -Naur mdadm-3.2.2/mdmon.8 mdadm-6218489/mdmon.8
--- mdadm-3.2.2/mdmon.8	2011-06-17 07:15:50.000000000 +0200
+++ mdadm-6218489/mdmon.8	2011-08-02 05:36:08.000000000 +0200
@@ -104,7 +104,7 @@
 relationship between a set of block devices and a raid array.  For
 example to create 2 arrays at different raid levels on a single
 set of disks, MD metadata requires the disks be partitioned and then
-each array can created be created with a subset of those partitions.  The
+each array can be created with a subset of those partitions.  The
 supported external formats perform this disk carving internally.
 .P
 Container devices simply hold references to all member disks and allow
@@ -172,7 +172,7 @@
 is automatically started by
 .I mdadm
 when needed and so does not need to be considered when working with
-RAID arrays.  The only times it is run other that by
+RAID arrays.  The only times it is run other than by
 .I  mdadm
 is when the boot scripts need to restart it after mounting the new
 root filesystem.
diff -Naur mdadm-3.2.2/mdstat.c mdadm-6218489/mdstat.c
--- mdadm-3.2.2/mdstat.c	2011-04-24 14:23:10.000000000 +0200
+++ mdadm-6218489/mdstat.c	2011-08-02 05:36:08.000000000 +0200
@@ -257,10 +257,10 @@
 				if (strncmp(w, "check", 5)==0)
 					ent->resync = 3;
 
-				if (l > 8 && strcmp(w+l-8, "=DELAYED"))
-					ent->percent = 0;
-				if (l > 8 && strcmp(w+l-8, "=PENDING"))
-					ent->percent = 0;
+				if (l > 8 && strcmp(w+l-8, "=DELAYED") == 0)
+					ent->percent = PROCESS_DELAYED;
+				if (l > 8 && strcmp(w+l-8, "=PENDING") == 0)
+					ent->percent = PROCESS_PENDING;
 			} else if (ent->percent == -1 &&
 				   w[0] >= '0' &&
 				   w[0] <= '9' &&
diff -Naur mdadm-3.2.2/platform-intel.h mdadm-6218489/platform-intel.h
--- mdadm-3.2.2/platform-intel.h	2011-04-24 14:23:11.000000000 +0200
+++ mdadm-6218489/platform-intel.h	2011-08-02 05:36:08.000000000 +0200
@@ -167,21 +167,6 @@
 	return r;
 }
 
-/**
- * imsm_orom_default_chunk - return the largest chunk size supported via orom
- * @orom: orom pointer from find_imsm_orom
- */
-static inline int imsm_orom_default_chunk(const struct imsm_orom *orom)
-{
-	int fs = fls(orom->sss);
-
-	if (!fs)
-		return 0;
-
-	return min(512, (1 << fs));
-}
-
-
 enum sys_dev_type {
 	SYS_DEV_UNKNOWN = 0,
 	SYS_DEV_SAS,
@@ -189,7 +174,6 @@
 	SYS_DEV_MAX
 };
 
-
 struct sys_dev {
 	enum sys_dev_type type;
 	char *path;
diff -Naur mdadm-3.2.2/super-intel.c mdadm-6218489/super-intel.c
--- mdadm-3.2.2/super-intel.c	2011-06-17 07:15:50.000000000 +0200
+++ mdadm-6218489/super-intel.c	2011-08-02 05:36:08.000000000 +0200
@@ -341,7 +341,7 @@
 		struct extent *e; /* for determining freespace @ create */
 		int raiddisk; /* slot to fill in autolayout */
 		enum action action;
-	} *disks;
+	} *disks, *current_disk;
 	struct dl *disk_mgmt_list; /* list of disks to add/remove while mdmon
 				      active */
 	struct dl *missing; /* disks removed while we weren't looking */
@@ -1518,9 +1518,9 @@
 			fd2devname(fd, buf);
 			printf("          Port%d : %s", port, buf);
 			if (imsm_read_serial(fd, NULL, (__u8 *) buf) == 0)
-				printf(" (%s)\n", buf);
+				printf(" (%.*s)\n", MAX_RAID_SERIAL_LEN, buf);
 			else
-				printf("()\n");
+				printf(" ()\n");
 		}
 		close(fd);
 		free(path);
@@ -2201,7 +2201,7 @@
 	if (prev_map)
 		map_to_analyse = prev_map;
 
-	dl = super->disks;
+	dl = super->current_disk;
 
 	info->container_member	  = super->current_vol;
 	info->array.raid_disks    = map->num_members;
@@ -2263,11 +2263,13 @@
 		info->new_chunk = info->array.chunk_size;
 		info->delta_disks = 0;
 	}
-	info->disk.major = 0;
-	info->disk.minor = 0;
+
 	if (dl) {
 		info->disk.major = dl->major;
 		info->disk.minor = dl->minor;
+		info->disk.number = dl->index;
+		info->disk.raid_disk = get_imsm_disk_slot(map_to_analyse,
+							  dl->index);
 	}
 
 	info->data_offset	  = __le32_to_cpu(map_to_analyse->pba_of_lba0);
@@ -2326,7 +2328,9 @@
 
 			dprintf("IMSM: General Migration checkpoint : %llu "
 			       "(%llu) -> read reshape progress : %llu\n",
-				units, blocks_per_unit, info->reshape_progress);
+				(unsigned long long)units,
+				(unsigned long long)blocks_per_unit,
+				info->reshape_progress);
 
 			used_disks = imsm_num_data_members(dev, 1);
 			if (used_disks > 0) {
@@ -4327,7 +4331,7 @@
 			devname);
 		return 1;
 	}
-	set_imsm_ord_tbl_ent(map, dk->number, dl->index);
+	set_imsm_ord_tbl_ent(map, dk->raid_disk, dl->index);
 	dl->disk.status = CONFIGURED_DISK;
 
 	/* if we are creating the first raid device update the family number */
@@ -4347,7 +4351,7 @@
 		mpb->family_num = __cpu_to_le32(sum);
 		mpb->orig_family_num = mpb->family_num;
 	}
-
+	super->current_disk = dl;
 	return 0;
 }
 
@@ -4912,6 +4916,15 @@
 	return 0;
 }
 
+static int imsm_default_chunk(const struct imsm_orom *orom)
+{
+	/* up to 512 if the plaform supports it, otherwise the platform max.
+	 * 128 if no platform detected
+	 */
+	int fs = max(7, orom ? fls(orom->sss) : 0);
+
+	return min(512, (1 << fs));
+}
 
 #define pr_vrb(fmt, arg...) (void) (verbose && fprintf(stderr, Name fmt, ##arg))
 /*
@@ -4940,15 +4953,16 @@
 			level, raiddisks, raiddisks > 1 ? "s" : "");
 		return 0;
 	}
-	if (super->orom && level != 1) {
-		if (chunk && (*chunk == 0 || *chunk == UnSet))
-			*chunk = imsm_orom_default_chunk(super->orom);
-		else if (chunk && !imsm_orom_has_chunk(super->orom, *chunk)) {
-			pr_vrb(": platform does not support a chunk size of: "
-			       "%d\n", *chunk);
-			return 0;
-		}
+
+	if (chunk && (*chunk == 0 || *chunk == UnSet))
+		*chunk = imsm_default_chunk(super->orom);
+
+	if (super->orom && chunk && !imsm_orom_has_chunk(super->orom, *chunk)) {
+		pr_vrb(": platform does not support a chunk size of: "
+		       "%d\n", *chunk);
+		return 0;
 	}
+
 	if (layout != imsm_level_to_layout(level)) {
 		if (level == 5)
 			pr_vrb(": imsm raid 5 only supports the left-asymmetric layout\n");
@@ -5298,9 +5312,8 @@
 	if (level && layout && *layout == UnSet)
 		*layout = imsm_level_to_layout(*level);
 
-	if (chunk && (*chunk == UnSet || *chunk == 0) && 
-	    super && super->orom)
-		*chunk = imsm_orom_default_chunk(super->orom);
+	if (chunk && (*chunk == UnSet || *chunk == 0))
+		*chunk = imsm_default_chunk(super->orom);
 }
 
 static void handle_missing(struct intel_super *super, struct imsm_dev *dev);
@@ -8661,8 +8674,9 @@
 		dprintf("imsm: info: Volume operation\n");
 		/* find requested device */
 		while (dev) {
-			imsm_find_array_minor_by_subdev(dev->index, st->container_dev, &devnum);
-			if (devnum == geo.dev_id)
+			if (imsm_find_array_minor_by_subdev(
+				    dev->index, st->container_dev, &devnum) == 0
+			    && devnum == geo.dev_id)
 				break;
 			dev = dev->next;
 		}
diff -Naur mdadm-3.2.2/super1.c mdadm-6218489/super1.c
--- mdadm-3.2.2/super1.c	2011-06-17 07:15:50.000000000 +0200
+++ mdadm-6218489/super1.c	2011-08-02 05:36:08.000000000 +0200
@@ -313,7 +313,7 @@
 		printf("\n");
 	}
 	if (sb->devflags) {
-		printf("      Flags :");
+		printf("          Flags :");
 		if (sb->devflags & WriteMostly1)
 			printf(" write-mostly");
 		printf("\n");
@@ -1056,6 +1056,8 @@
 		sb->dev_number = __cpu_to_le32(di->disk.number);
 		if (di->disk.state & (1<<MD_DISK_WRITEMOSTLY))
 			sb->devflags |= __cpu_to_le32(WriteMostly1);
+		else
+			sb->devflags &= ~(__cpu_to_le32(WriteMostly1));
 
 		if ((rfd = open("/dev/urandom", O_RDONLY)) < 0 ||
 		    read(rfd, sb->device_uuid, 16) != 16) {
diff -Naur mdadm-3.2.2/sysfs.c mdadm-6218489/sysfs.c
--- mdadm-3.2.2/sysfs.c	2011-06-17 07:15:50.000000000 +0200
+++ mdadm-6218489/sysfs.c	2011-08-02 05:36:08.000000000 +0200
@@ -470,7 +470,7 @@
 	lseek(fd, 0, 0);
 	n = read(fd, buf, sizeof(buf));
 	if (n <= 0)
-		return -1;
+		return -2;
 	buf[n] = 0;
 	*val = strtoull(buf, &ep, 0);
 	if (ep == buf || (*ep != 0 && *ep != '\n' && *ep != ' '))
diff -Naur mdadm-3.2.2/util.c mdadm-6218489/util.c
--- mdadm-3.2.2/util.c	2011-06-17 07:15:50.000000000 +0200
+++ mdadm-6218489/util.c	2011-08-02 05:36:08.000000000 +0200
@@ -146,16 +146,16 @@
 {
 	struct utsname name;
 	char *cp;
-	int a,b,c;
+	int a = 0, b = 0,c = 0;
 	if (uname(&name) <0)
 		return -1;
 
 	cp = name.release;
 	a = strtoul(cp, &cp, 10);
-	if (*cp != '.') return -1;
-	b = strtoul(cp+1, &cp, 10);
-	if (*cp != '.') return -1;
-	c = strtoul(cp+1, NULL, 10);
+	if (*cp == '.')
+		b = strtoul(cp+1, &cp, 10);
+	if (*cp == '.')
+		c = strtoul(cp+1, &cp, 10);
 
 	return (a*1000000)+(b*1000)+c;
 }
@@ -535,6 +535,7 @@
 	struct supertype *st = guess_super(fd);
 
 	if (!st) return 0;
+	st->ignore_hw_compat = 1;
 	st->ss->load_super(st, fd, name);
 	/* Looks like a raid array .. */
 	fprintf(stderr, Name ": %s appears to be part of a raid array:\n",
