--- a/utils/gssd/context_lucid.c	2012-07-03 10:07:32.742358732 +0200
+++ b/utils/gssd/context_lucid.c	2012-07-03 10:15:18.461202054 +0200
@@ -266,7 +266,7 @@
 	int retcode = 0;
 
 	printerr(2, "DEBUG: %s: lucid version!\n", __FUNCTION__);
-	maj_stat = gss_export_lucid_sec_context(&min_stat, &ctx,
+	maj_stat = gss_krb5_export_lucid_sec_context(&min_stat, &ctx,
 						1, &return_ctx);
 	if (maj_stat != GSS_S_COMPLETE) {
 		pgsserr("gss_export_lucid_sec_context",
@@ -302,7 +302,7 @@
 	else
 		retcode = prepare_krb5_rfc4121_buffer(lctx, buf, endtime);
 
-	maj_stat = gss_free_lucid_sec_context(&min_stat, ctx, return_ctx);
+	maj_stat = gss_krb5_free_lucid_sec_context(&min_stat, ctx);
 	if (maj_stat != GSS_S_COMPLETE) {
 		pgsserr("gss_free_lucid_sec_context",
 			maj_stat, min_stat, &krb5oid);
--- a/utils/gssd/krb5_util.c	2012-07-03 10:12:27.984734761 +0200
+++ b/utils/gssd/krb5_util.c	2012-07-03 10:13:00.097449211 +0200
@@ -1347,10 +1347,10 @@
 	 * list of supported enctypes, use local default here.
 	 */
 	if (krb5_enctypes == NULL || limit_to_legacy_enctypes)
-		maj_stat = gss_set_allowable_enctypes(&min_stat, credh,
+		maj_stat = gss_krb5_set_allowable_enctypes(&min_stat, credh,
 					&krb5oid, num_enctypes, enctypes);
 	else
-		maj_stat = gss_set_allowable_enctypes(&min_stat, credh,
+		maj_stat = gss_krb5_set_allowable_enctypes(&min_stat, credh,
 					&krb5oid, num_krb5_enctypes, krb5_enctypes);
 
 	if (maj_stat != GSS_S_COMPLETE) {
--- a/utils/gssd/svcgssd_krb5.c	2012-07-03 10:16:40.876439303 +0200
+++ b/utils/gssd/svcgssd_krb5.c	2012-07-03 10:28:25.168934845 +0200
@@ -210,7 +210,7 @@
 			"enctypes from defaults\n", __func__, num_enctypes);
 	}
 
-	maj_stat = gss_set_allowable_enctypes(&min_stat, gssd_creds,
+	maj_stat = gss_krb5_set_allowable_enctypes(&min_stat, gssd_creds,
 			&krb5oid, num_enctypes, enctypes);
 	if (maj_stat != GSS_S_COMPLETE) {
 		printerr(1, "WARNING: gss_set_allowable_enctypes failed\n");
