--- a/saslauthd/auth_krb5.c	2012-07-11 00:16:09.968102908 +0200
+++ b/saslauthd/auth_krb5.c	2012-07-11 00:22:56.996962718 +0200
@@ -418,11 +418,13 @@
 	return strdup("NO saslauthd internal error");
     }
     
+#ifdef AUTH_KRB5_VERIFY_TGT
     if (!k5support_verify_tgt(context, ccache)) {
 	syslog(LOG_ERR, "auth_krb5: k5support_verify_tgt");
 	result = strdup("NO saslauthd internal error");
 	goto fini;
     }
+#endif
     
     /* 
      * fall through -- user is valid beyond this point  
